### ISSUE TITLE
Getting error on publishing text of length greater than around 35 characters.

### ISSUE THREAD

**Issue by abhinavzspace** [slact/nchan#332]
When I am publishing some text data through a websocket to a publisher endpoint and the text data is greater than 35 characters, I get this error:

```
Assertion failed: (fastlen % __vector_size_bytes == 0), function websocket_unmask_frame, file /usr/local/share/nchan-nginx-module/src/subscribers/websocket.c, line 311.
2017/03/23 12:54:55 [alert] 7756#0: worker process 7757 exited on signal 6
```

Occasionally this also:

```
2017/03/23 13:07:35 [error] 8079#0: MEMSTORE:01: tried adding WAITING chanhead 00007F885B4044B0 /xyz_abhinavzspace to chanhead_gc. why?
```

My nginx config is:

```
worker_processes 5;

worker_rlimit_nofile 1024;

events {
    worker_connections  1024;
}

http {

  server {
    listen       8090;

    location ~ /pubsub/([\w+|-]+)/(\w+)$ { # browser - $1 is noteId, $2 is username
      nchan_pubsub;
      nchan_channel_id "$1_$2";

      nchan_publisher_upstream_request /upstream_pub;
    }

    location ~ /pub/([\w+|-]+)/(\w+)$ { # server - $1 is noteId, $2 is username
      nchan_publisher;
      nchan_channel_id "$1_$2";
    }

    location = /upstream_pub {
      proxy_pass http://127.0.0.1:5555/from_client;
      proxy_set_header X-Publisher-Type $nchan_publisher_type;
      proxy_set_header X-Prev-Message-Id $nchan_prev_message_id;
      proxy_set_header X-Channel-Id $nchan_channel_id;
      proxy_set_header X-Original-URI $request_uri;
    }

  }
}
```

Subscribing on `http://127.0.0.1:8090/pubsub/xyz/abc` through websocket works but publishing is failing.

Its so strange. When i do `websocket.send(<string of 40 characters>)` it does not work but when I do `websocket.send(<string of 30 characters>)`, it works.

Do let me know if you need any more details.

(This is my first time with nginx and nchan. So if I am doing something completely stupid, I apologise)

Thanks

**Comment by slact** [comment]
Can you show me the output of `cat /proc/cpuinfo` ?

**Comment by abhinavzspace** [comment]
uname -a    =>    `Darwin abhinavs-MacBook-Pro.local 15.6.0 Darwin Kernel Version 15.6.0: Wed Nov  2 20:30:56 PDT 2016; root:xnu-3248.60.11.1.2~2/RELEASE_X86_64 x86_64`

cat /proc/cpuinfo    =>    `No such file or directory`

I installed through    =>    `brew tap homebrew/nginx; brew install nginx-full --with-nchan-module `

**Comment by slact** [comment]
Thanks. On a macbook, I need the output of `sysctl -n machdep.cpu.brand_string` to get CPU info.

**Comment by abhinavzspace** [comment]
Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz

**Comment by slact** [comment]
This is a bug in Nchan's AVX2 websocket optimizations. I'll have it fixed shortly.

**Comment by slact** [comment]
I'm having trouble reproducing this, so i'm going to need your help fixing it.
Please do the following:

```
brew uninstall nginx-full
brew untap homebrew/nginx
brew tap slact/homebrew-nginx
brew install nginx-full --with-nchan-module
```

then edit your nginx.conf to add an error_log line in the beginning:
```nginx
worker_processes 5;

error_log '/pat/to/error.log' notice;

worker_rlimit_nofile 1024;

events {
    worker_connections  1024;
}

http {
   #...
}
```

Run nginx and send a websocket message. When the worker crashes, you should have a like in the error log that looks like:
```
__vector_size_bytes: n, payload_len: n, preamble_len: n, i: n, fastlen: n
```
That's the line I need to see.

**Comment by abhinavzspace** [comment]
```
2017/03/24 08:35:22 [error] 19728#0: SUB:WEBSOCKET:__vector_size_bytes: 32, payload_len: 40, preamble_len: 16, i: 16, fastlen: 16
Assertion failed: (0), function websocket_unmask_frame, file /usr/local/share/nchan-nginx-module/src/subscribers/websocket.c, line 314.
```

**Comment by abhinavzspace** [comment]
Do let me know if anything else needed from my side. Thank you.

**Comment by slact** [comment]
Thanks. I've identified the bug and will have a fix shortly.

**Comment by slact** [comment]
alright, it should be fixed. please update your slact/homebrew-nginx tap and rebuild nginx-full. Once you have verified that it works correctly I'll tag and release a new version.

**Comment by abhinavzspace** [comment]
I am getting 
```
abhinavs-MBP:~ abhinav$ brew install nginx-full --with-nchan-module
==> Installing nginx-full from slact/nginx
Error: nchan-nginx-module: /usr/local/Homebrew/Library/Taps/slact/homebrew-nginx/Formula/nchan-nginx-module.rb:4: syntax error, unexpected <<, expecting keyword_end
<<<<<<< HEAD
  ^
/usr/local/Homebrew/Library/Taps/slact/homebrew-nginx/Formula/nchan-nginx-module.rb:7: syntax error, unexpected ===, expecting keyword_end
=======
   ^
/usr/local/Homebrew/Library/Taps/slact/homebrew-nginx/Formula/nchan-nginx-module.rb:10: syntax error, unexpected >>, expecting keyword_end
>>>>>>> nchan ws-optimize-debug
```

Is merging done correctly? Or am I not installing it correctly?

I followed these steps:
```
brew uninstall nginx-full
brew untap homebrew/nginx
brew tap slact/homebrew-nginx
brew install nginx-full --with-nchan-module
```

**Comment by abhinavzspace** [comment]
I am not able to successfully install the latest fixed version. I checked `/usr/local/share/nchan-nginx-module/src/subscribers/websocket.c`. It is old file.

I cant figure out the steps to update. :(

**Comment by slact** [comment]
Try this:
```
brew uninstall nginx-full
brew untap slact/homebrew-nginx
brew tap slact/homebrew-nginx
brew install --HEAD nginx-full --with-nchan-module
```

**Comment by abhinavzspace** [comment]
I double checked. It works.
Great!

**Comment by slact** [comment]
Ok, i've released version 1.1.3 with this fix. Thanks for reporting it. you can go back to the `homebrew/nginx` tap:
```
brew uninstall nginx-full
brew untap slact/homebrew-nginx
brew tap homebrew/nginx
brew install nginx-full --with-nchan-module
```

Thanks for reporting the issue and helping me debug it.
